File: LC145.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        
        vector<int> v;
        
        if(root == NULL)
            return v;

        stack<TreeNode *> st;
        TreeNode *t = root;

        while(t or !st.empty()){
            if(!t){
                t = st.top();
                st.pop();
            }
            else{
                if(t->left)
                    st.push(t->left);
            v.push_back(t->val);
            t = t->right;
            }
        }
        
        reverse(v.begin(),v.end());
        return v;
    }
};

//Recursive solution
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void postOrder(vector<int> &v, TreeNode *root){
        if(root == NULL)
            return;
        
        postOrder(v,root->left);
        postOrder(v,root->right);
        v.push_back(root->val);
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> v;
        
        postOrder(v,root);
        
        return v;
    }
};

File: LC145.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        st = []
        l = []
        
        if root == None:
            return l

        t = root

        while t or len(st) != 0:
            if not t:
                t = st[-1]
                st.pop()
            else:
                if t.left:
                    st.append(t.left)
                l.append(t.val)
                t = t.right
        
        l.reverse()
        return l
